Define custom completion for VideoModule

jcdyer opened the pull request:
Update the VideoModule to publish a completion event when the player
reaches 95% complete, and submit a BlockCompletion when that event
occurs.

## Testing instructions

1.  In the lms shell, remove all existing BlockCompletions for video blocks.
    ```python
    from lms.djangoapps.completion.models import BlockCompletion
    BlockCompletion.objects.filter(block_type='video').delete()
    ```
2.  Sign in to the LMS and visit the demo course.
3.  Navigate to "Example Week 1: Getting Started > Lesson 1 - Getting Started > (Vertical 2) Working with Videos".  Open the browser's development panel, and view network traffic.  
4. Verify that when the video reach 4m 3s, a POST is sent to a URL for the block ending with "publish_completion", and receives a 200 response.  
5.  Verify that a BlockCompletion now exists for the video block with completion value == 1.0.
    ```python
    from lms.djangoapps.completion.models import BlockCompletion
    print(BlockCompletion.objects.filter(block_type='video', completion=1.0))
    ```
6.  Navigate to the next vertical: "Videos on edX".  This one uses more features (a custom start time and end-time).  Watch to the end of the video (or jump to a few seconds before the end of the video, and just watch the end). 
7.  Verify that a publish_completion POST is sent after 7m 17s (Start time is 5m 10s, end time is 7m 24s = 134s; 134 * .95 = 127.3s; 5m 10s + 127s = 7m 17s), and receives a 200 response.
8.  Verify that a BlockCompletion object exists for this block as well.

## Author concerns

I'm rusty at Javascript, and entirely new to the video player.  I may not be doing things the right way. 

## Reviewers
- [x] @pomegranited 
- [x] @robrap 

FYI: @yro for video player expertise.

## TODO:

- [x] Squash commits
- [x] Rebase PR against edx:master after cliff/complete-default-blocks merges ( #16234 )

edx-webhook commented:
Thanks for the pull request, @jcdyer! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=16569"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

jcdyer commented:
jenkins run python
jenkins run quality

jcdyer commented:
jenkins run quality

jcdyer commented:
jenkins run python

jcdyer commented:
jenkins run quality

jzoldak commented:
jenkins run quality

pomegranited commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,123 @@
+                this.startTime = this.state.config.startTime;
+                this.endTime = this.state.config.endTime;
+                if (this.endTime) {
+                    this.completeAfter = this.startTime + (this.endTime - this.startTime) * this.completionPercentage;
```
This calculation is the same as [below](https://github.com/edx/edx-platform/pull/16569/files#diff-6ec38f7c3b767f48055a8c9b915744f9R78) -- can it be moved to a separate method, e.g. `this.calculateCompleteAfter(duration)`.

Could even include the `if (duration == 0)` check from [here](https://github.com/edx/edx-platform/pull/16569/files#diff-6ec38f7c3b767f48055a8c9b915744f9R72), to ensure that `this.endTime - this.startTime > 0`?

jcdyer replied:
I think I'll leave the duration check outside the method, for a few reasons: 

1. `duration === 0` is a special case where the data doesn't exist yet (should probably be returning undefined, but I don't want to mess with it), which is the only case that really needs to be checked.  
2. If the video is actually short, it won't hurt to submit the completion anyway.  Worst case, the `ended` event occurs at the same time, and it sends two POSTs instead of one, but the DB is protected from this by a unique index.
3. The check for a zero length video should actually have been `if (duration === this.startTime)`, not `if (duration === 0)`.  That's only true for videos that don't have a custom end time.

pomegranited commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,123 @@
+                    if (self.complete) {
+                        return;
+                    }
+                    if (self.lastSentAt !== undefined && currentTime - self.lastSentAt < 3.0) {
```
Could you pull that `3.0` into a variable at the top, so we can easily adjust if needed?

pomegranited commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,123 @@
+                        type: 'POST',
+                        url: this.state.config.publishCompletionUrl,
+                        contentType: 'application/json',
+                        data: JSON.stringify({completion: 1.0})
```
Cleaner way to post JSON data in [035_video_accessible_menu.js](https://github.com/open-craft/edx-platform/blob/1b9342faaa0ca8c0c6f118e70cc56c0ee10f2699/common/lib/xmodule/xmodule/js/src/video/035_video_accessible_menu.js#L52-L63)

jcdyer replied:
When I try that (`data: {completion: 1.0}, dataType: 'json'`), it submits the data as x-www-form-encoded, and returns a 400 error.  I've changed it back, but left in a commit (https://github.com/edx/edx-platform/pull/16569/commits/a0f7ed57a14de2b30c539de9638801b65c571e20) that tries to do it the way 035_video_accessible_menu does it the way it's done in 035_video_accessible_menu.js, in case you can see something obvious I was missing.

pomegranited commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,123 @@
+                this.el.remove();
+                this.state.el.off('destroy', this.destroy);
+                delete this.state.completionHandler;
+            },
```
This is all a bit mysterious to me, but I can see there's [other JS libraries doing something similar|https://github.com/open-craft/edx-platform/blob/1b9342faaa0ca8c0c6f118e70cc56c0ee10f2699/common/lib/xmodule/xmodule/js/src/video/09_events_plugin.js#L34-L37], so 👍

jcdyer replied:
Yeah.  I admit this part was copypasta from one of the other scripts.  I think it's just making sure that when the player gets destroyed (maybe when navigating away from the vertical?) all the resources get cleaned up appropriately--no dangling references causing memory leaks.  But I didn't track it all down.

pomegranited commented on common/lib/xmodule/xmodule/video_module/video_handlers.py:
```diff
@@ -201,6 +202,28 @@ def get_static_transcript(self, request, transcripts):
+        """
+        completion_service = self.runtime.service(self, 'completion')
+        if not completion_service or not completion_service.completion_tracking_enabled():
+            raise JsonHandlerError(404, u"Completion service not available")
```
My only problem with this is that the console logs get filled with 500 errors if the completion tracking waffle switch isn't set.  Shouldn't that case fail quietly?

jcdyer replied:
New version passes completionEnabled to the client, and the client only attaches the event handlers if completion is enabled.

robrap replied:
Thanks for splitting these.  I still think changing the second from "Completion service not available" to something like "Completion tracking is not enabled and api calls are unexpected" would help someone understand more quickly where to look for the problem.

pomegranited commented on common/lib/xmodule/xmodule/video_module/video_handlers.py:
```diff
@@ -281,6 +304,8 @@ def transcript(self, request, dispatch):
                     transcripts, transcript_format=self.transcript_download_format, lang=lang
                 )
+            except (KeyError, UnicodeDecodeError):
+                return Response(status=404)
```
This looks unrelated?

jcdyer replied:
I moved it up from below while debugging something because it turns out UnicodeDecodeError is a subclass of ValueError, so that case was getting swallowed by the other `except` handler block below.

pomegranited commented on cms/envs/common.py:
```diff
@@ -1451,3 +1451,6 @@
 VIDEO_IMAGE_ASPECT_RATIO_ERROR_MARGIN = 0.1
+
+
+COMPLETION_VIDEO_COMPLETE_PERCENTAGE = 0.95
```
Also, could you add it to `[cl]ms.envs.aws` too, so we can override it with environment settings as required?

pomegranited commented on lms/djangoapps/courseware/tests/test_video_mongo.py:
```diff
@@ -198,6 +194,7 @@ class TestGetHtmlMethod(BaseTestXmodule):
     CATEGORY = "video"
     DATA = SOURCE_XML
     METADATA = {}
+    maxDiff = None
```
Artefact of test debugging -- please remove.

pomegranited commented on lms/djangoapps/courseware/tests/test_video_mongo.py:
```diff
@@ -231,8 +228,17 @@ def setUp(self):
+            'completionPercentage': 0.95,
+            'publishCompletionUrl': self.item_descriptor.xmodule_runtime.handler_url(
+                self.item_descriptor, 'publish_completion', ''
+            ).rstrip('?'),
```
Can use `get_handler_url` method defined below too?

jcdyer commented:
jenkins run all

jcdyer commented:
jenkins run bokchoy

jcdyer commented:
jenkins run lettuce

jcdyer commented:
jenkins run all

jcdyer commented:
This is ready for another review.

pomegranited commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,142 @@
+                        url: this.state.config.publishCompletionUrl,
+                        contentType: 'application/json',
+                        dataType: 'json',
+                        data: JSON.stringify({completion: 1.0}),
```
@jcdyer I messed around with a couple of the [jquery.ajax options for posting data](https://api.jquery.com/jQuery.ajax/#sending-data-to-server), like `processData: false` and different options for `contentType`, but had no luck getting the data parsed properly by the `XBlock.json_handler`.

Also found this [stackoverflow post](https://stackoverflow.com/a/6323528/4302112) which suggests stringifying the data like you did.

So no worries here :)

pomegranited commented:
:+1: @jcdyer 

I left one nit about a comment, but it's working well.  Thanks for the updates!

- [x] I tested this on my devstack using the PR instructions, and with and without the waffle flag described in https://github.com/edx/edx-platform/pull/16234.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation -- docstrings only here; [asked about waffle flag documentation](https://github.com/edx/edx-platform/pull/16234#issuecomment-345576589)

jcdyer commented:
@robrap, this is squashed and ready for an edx review.

jcdyer commented:
@robrap Do you have a sense of timeline for when you (or someone else at edx) would be able to review this?

robrap commented:
@jcdyer: Sorry for the delay.  I was trying to clarify the reviewer at edX, and in doing that, I forgot that there was one waiting.  Feel free to ping me on HipChat if you don't see action in 1-3 days, even if it is just to get a schedule.

In any case, at this time, I am officially going to review the rest of the block completion PRs.  The reviewer for the aggregate PRs is still a question.

For this review, I can get to it in the next couple of days.  I'm just wrapping up a story and then I'll pick this up.

Thanks for checking back in.

robrap commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,142 @@
+(function(define) {
```
Seems like there should be tests added to:
https://github.com/edx/edx-platform/tree/master/common/lib/xmodule/xmodule/js/spec/video

robrap replied:
Does thumbs-up mean this is coming?

jcdyer replied:
Yes.  I worked on tests on Friday, but wasn't able to get them all in place. I'll ping you again when they are ready or by end of day at the latest.

jcdyer commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,142 @@
+    define('video/09_completion.js', [], function() {
+    /**
+     * Completion handler
+     * @exports video/09_completion.js
```
By looking at the other files in video.  They seem to be roughly ordered by dependency.  completion is required by 10_main, so it needed to be less than 10, but I figured not much else depends on it, so it can be added late in the process.

I didn't really understand why the numbering is needed here, but I figured it was better to be consistent with the subsystem I'm working in.

jcdyer commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,142 @@
+                }
+
+                if (currentTime > this.completeAfter) {
+                    this.lastSentAt = currentTime;
```
> Seems like lastSentAt would be set in markCompletion

I can be down with that.  I'll make the update.

> On a related note, would handleEnded need the same throttling, or is that a single event?

handleEnded is just a single event, so I thought it would be better to keep it simle.

> If handleEnded ajax call fails is the event repeated?

No.  That could be useful, but it brings up a few new questions.  What if the server's down for five minutes?  Or if the user is on a train, and the wifi went out for half an hour?  How often and how long should we repeat?  Should we use local storage to keep stale completions around until the user is back online?  How do we determine when they come back online?  

Ultimately, I think the 100% solution is going to be a much bigger one, and I'm not sure we need to hit 100% (because as you suggest, the user can go back and recomplete it fairly easily).  The question is how good is good enough?  Is a single submission sufficient?  Should we retry three times, every five seconds, and call it quits after that?

> If someone fasts forwards to the end of a video, does that make it seem like it was watched?
Basically, just want to make sure someone can easily mark it as complete if there is a temporary problem.

Yes.  As long as the user is watching a portion of video after the 95% mark when timeupdate gets triggered, the video is considered complete.

jcdyer commented on common/lib/xmodule/xmodule/video_module/bumper_utils.py:
```diff
@@ -137,6 +137,9 @@ def bumper_metadata(video, sources):
             video.runtime.handler_url(video, 'transcript', 'available_translations').rstrip('/?'), 'is_bumper', 1
         ),
+        'publishCompletionUrl': set_query_parameter(
+            video.runtime.handler_url(video, 'publish_completion', '').rstrip('?'), 'is_bumper', 1
```
Yeah.  I tested both ways, and noticed that the trailing / wasn't actually changing the behavior of the call, so I figured why worry about it?

jcdyer commented on common/lib/xmodule/xmodule/video_module/video_handlers.py:
```diff
@@ -201,6 +202,28 @@ def get_static_transcript(self, request, transcripts):
+        completion_service = self.runtime.service(self, 'completion')
+        if not completion_service or not completion_service.completion_tracking_enabled():
+            raise JsonHandlerError(404, u"Completion service not available")
+        if not 0.0 <= data['completion'] <= 1.0:
```
Good idea.  In python 2, a non-number will just sort outside the range of numbers, but in python 3 it will become an error.  Better not to introduce a possible 500 here.

robrap commented:
@jcdyer: What's the status of this?  I'd like to close my review ticket, but I haven't seen you ping me for re-review.  Thanks.

jcdyer commented:
Hi Robert, sorry for the delay. I've been off this week. I will see if I can get it wrapped up and back on your plate today.

On December 12, 2017 3:43:45 PM EST, Robert Raposa <notifications@github.com> wrote:
>@jcdyer: What's the status of this?  I'd like to close my review
>ticket, but I haven't seen you ping me for re-review.  Thanks.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/edx/edx-platform/pull/16569#issuecomment-351188165

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

jcdyer commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,142 @@
+(function(define) {
+    'use strict';
+    define('video/09_completion.js', [], function() {
+    /**
```
Updating throughout.

jcdyer commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,142 @@
+                        // or fall back to 'ended' handler.
+                        return;
+                    }
+                    this.completeAfter = this.calculateCompleteAfter(this.startTime, duration);
```
Updated.

robrap commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,144 @@
+                    this.completeAfterSeconds = this.calculateCompleteAfterSeconds(this.startTime, duration);
+                }
+
+                if (currentTime > this.completeAfterSeconds) {
```
I had proposed `completeAfter` => `completeAfterTime`.  You instead went with `completeAfterSeconds`.

I find the code confusing. Since `currentTime` and `this.completeAfterSeconds`, they seem to contain the same unit.  I'd like to see then consistently named.  If they are both times, then please rename the latter to `this.completeAfterTime`.  If they are both in seconds from the beginning, then options might be:
- rename all the "xxxTime" variables to "xxxSeconds", or 
- drop a comment where they are defined, and just explain what they contain.

Or if I am missing something entirely, feel free to help clarify and make some other code change that makes this comparison more clear.  Thanks.

jcdyer replied:
@robrap I was thinking that time and seconds were different, analogous to the way datetime and timedelta are different, but it looks like you're right, and Javascript treats them identically.  I think you're right that they should all have the same suffix.  I'm leaning toward going with `xxxSeconds`, even though the names are slightly more awkward, because it makes the semantics clearer in all contexts (including the Javascript code, the Django settings modules, and JSON files that feed into the settings).

robrap replied:
@jcdyer What does seconds actually mean?  What are sample values for startTime (or startSeconds)?  Maybe you could clarify this with a quick comment if the name isn't enough to understand.

jcdyer replied:
On further look, I've settled on calling everything "Time."  I was hoping to distinguish between DateTime-like objects (which in javascript are represented as "seconds since the epoch", and the number of seconds from the beginning of a video, but even that distinction was going to end up muddied, because the video player exposes "startTime" and "endTime" (which are seconds since the start of the video), so as per your initial recommendation, I'm settling on a consistent use of "Time," with a comment to explain where meaning might be confusing.

jcdyer commented:
@robrap Tests are added, and I've addressed your last comments (about naming and error messaging).  If the build is green, go ahead and review.  JS tests and quality passed locally.  I'll check in the morning to be sure Jenkins is happy.

jcdyer commented on common/lib/xmodule/xmodule/js/spec/video/html5_video_spec.js:
```diff
@@ -220,7 +220,7 @@
                     });
 
-                    it('set new inccorrect values', function() {
+                    it('set new incorrect values', function() {
```
This test seems to be flaky.  I saw it fail once, and couldn't bear to let the typo survive.  I'm not sure what the process is for external contributors marking tests flaky, so I haven't created any tickets about it.

robrap replied:
If you know it to be flaky, I don't see any issue with you following the updated flaky test process.

jcdyer commented:
There was an error (wher I hadn't changed completeAfterSeconds to xxxTime in the specfile, so I cleaned it up, and added some documentation.  In adding the documentation, I noticed that I was registering the destroy handler at the wrong time, so I updated that, too.

robrap commented on common/lib/xmodule/xmodule/js/spec/video/completion_spec.js:
```diff
@@ -0,0 +1,55 @@
+
+        it('calls the completion api on the LMS when the time updates', function() {
+            spyOn(state.completionHandler, 'markCompletion').and.returnValue(null);
+            state.el.trigger('timeupdate', 24.0);
```
Consider adding another trigger and checking and changing the check for markCompletion to ensure it was called only once.

jcdyer replied:
Done.

robrap commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,173 @@
+             *  completion calculations.
+             */
+            initialize: function() {
+                this.lastSentTime = undefined;
```
This is where I had expected a comment.  Maybe you added them elsewhere?

I don't even know if my comment is accurate, but it would be nice to see something like:
```
// The variables that end with 'Time' are actually in seconds from 0, representing the actual start of the video.
```

jcdyer replied:
Done.

robrap replied:
Thanks.  The one that is confusing to me is what startTime represents.  Is that the startTime of real content, before which if you don't get there it won't be marked complete?  It's up to you whether not you enhance comments.  I'm just interested.

jcdyer replied:
You can actually define an edx video that only includes a subset of the source video, so if you have a 10 minute youtube video, you can include only the slice from 8m0s to 9m0s minute of the video by assigning your block a startTime of 480 (`8 * 60`), and an endTime of 540 (`9 * 60`).  So startTime is the time from the start of the source video that the video block will use as its start time.  There's a video in the demo course of a young man reading an excerpt from the Odyssey (or was it the Iliad?) that uses this feature, if you want to see it in action.

robrap replied:
Thanks for explaining that for me.  That makes more sense than what I was imagining.

robrap commented on common/lib/xmodule/xmodule/js/src/video/09_completion.js:
```diff
@@ -0,0 +1,173 @@
+            },
+
+            /** How many seconds to wait after a POST fails to try again. */
+            repostDelayTime: function() {
```
I actually would have left this as `repostDelaySeconds`, since this is not seconds from the beginning of the video like all the other Time variables.  I think it would be more consistent if this went back as it was.

jcdyer replied:
Makes sense.

jcdyer commented:
Python has apparently flaky paver tests, e.g. https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/46092/testReport/junit/pavelib.paver_tests.test_assets/TestPaverAssetTasks/test_compile_sass_4_____system_lms__/ (+13 others)

All other tests passed.  Rebasing against master to rerun tests.

jcdyer commented:
Python has apparently flaky paver tests, e.g. https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/46092/testReport/junit/pavelib.paver_tests.test_assets/TestPaverAssetTasks/test_compile_sass_4_____system_lms__/ (+13 others)

All other tests passed.  Rebasing against master to rerun tests.

robrap commented:
@jcdyer: I'm hopeful that they aren't flaky, but were just broken at one time.  Hopefully all will be well after rebase.

robrap commented:
@jcdyer: Was this a bad rebase, or do you need to repoint the PR against master or something?

jcdyer commented:
Oh wow.  I forgot this branch wasn't pointing at master.  The target branch merged to master on Nov 28th, so I will repoint the PR to master now.

robrap commented:
Also, feel free to squash whenever you are ready.  Thanks.

jcdyer commented:
Repointed, rebased, resolved conflicts, and squashed.
